### ISSUE TITLE
Revisit the version constraint for federated plugin implementations

### ISSUE THREAD

**Issue by amirh** [flutter/flutter#46264]
In https://flutter.dev/go/federated-plugins we planned to have the edge from the app-facing plugin to endorsed implementation have an "any" constraint.
Doing so fails `pub publish` checks, for now we're setting real constraints on that edge.

We should either update the pub client to allow this, or update the federated plugins design to have a real constraint there.

cc @jonasfj

**Comment by amirh** [comment]
The app-facing plugin (e.g url_launcher) does not use any public API provided by the platform implementation (url_launcher_web), thus as long as url_launcher_web works with the platform interface version it cannot break the app-facing plugin. This is why an upper bound on the url_launcher_web version that url_launcher depends on **is not needed**.

A platform implementation may expose a public API with platform-specific features used directly by the app (app developers that wants to use it will explicitly depend on the platform implementation with `^` constraint). This public API may have breaking changes and will result in major version bumps, however the breaking change does not affect the app-facing plugin, so an upper version bound for the url_launcher->url_launcher_web dependency **is not desired**.

@jonasfj I think we want to allow `any`, or lower bound only for app-facing -> platform-implementation dependencies.

**Comment by jonasfj** [comment]
Doesn't the generated code in the app rely on some API to obtain/register an instance of `url_launcher_platform_interface` from `url_launcher_web` package?

Also we don't know that if future versions of the `url_launcher_web` package will provide (or even depend on) the `url_launcher_platform_interface` package. I guess we could argue that if the package decided to implement a different platform interface then it fundamentally broke what it promised to be.

It's also worth noting that updating the version constraints and publishing a new `url_launcher` package is cheap to do (just do a `+1` version). It has low impact on downstream consumers as `pub get` will upgrade them, if this is necessary to resolve constraints (when upgrading/installing a different package).

On the other hand, retracting a broken package version is highly disruptive to downstream consumers. Indeed this quickly becomes infeasible once adoptions takes hold of a package version.

-------

If we decide that an `any` constraint is truly desirable, we might want to implement this as: `>= 0.0.0 <=9999.0.0`. That way there is no need for special casing in `pub`. It's certainly not clear how `pub` would recognize that a package is a platform-implementation, and therefore it's okay to depend on it without any constraints.

How would `pub` know if the package being published is depending on `url_launcher_web` because it's wrapping a platform specific feature, or because it's using the `url_launcher_platform_interface` exposed.

In the general case it is almost always a mistake to publish a package without constraints on its dependencies. I'm not entirely convinced that isn't also the case here.

/cc @sigurdm,

**Comment by stuartmorgan** [comment]
I think there's also the case of an internal breaking change in an implementation to consider (which would in turn be a breaking change for the app-facing plugin). As a concrete example that came up in discussion of `path_provider_windows` implementation, let's say we decide at some point that we need to change where `path_provider_windows` stores files, and for some reason couldn't do it as an opt-in and couldn't do migration (this may not be a realistic example, but I think it's plausible that a more realistic example like this could exist). That would be a breaking change in `path_provider_windows`, so the right way to express that seems like it would be to make a major version change in `path_provider_windows`, then make a major version change in `path_provider` and bump the `path_provider_windows` dependency to the new version. If we use `any` we can't do that.

So while it can't break the app facing plugin *itself*, a change in a platform implementation could be a conceptually breaking change for users.

**Comment by jonasfj** [comment]
Also to reiterate, if we have:
```yaml
name: url_launcher
version: 1.2.3
dependencies:
  url_launcher_web: ^1.0.0
...
```

And the author of `url_launcher_web` decides to release version `2.0.0`. Even though the version `2.0.0` still works for `url_launcher`, then it's easy to simply publish a new version of `url_launcher` allowing `>=1.0.0 <3.0.0`:

```yaml
name: url_launcher
version: 1.2.4 # new version number
dependencies:
  url_launcher_web: >=1.0.0 <3.0.0 # bump constraint from ^1.0.0, to allowing both 1.x and 2.x
...
```

----------------

When the author of `url_launcher_web` bumps major version, then it's probably best that developers of `url_launcher` reads the changelog before they claim to work with the newer version of `url_launcher_web`. And a minor version bump in `url_launcher` will have little impact on consumers who can easily pick it up.

If on the other hand, `url_launcher` had an `url_launcher_web: any` constraint, there would be no good remedies if 2.0.0 of `url_launcher_web` breaks `url_launcher`.

**Comment by stuartmorgan** [comment]
I just ran into another case of needing to enforce a minimum version: auto-registration was added to some of our implementations after the fact (once tool support was added), allowing us to remove the manual registration workaround from the main package. However, if we removed the manual registration without requiring at least the version that added the auto-registration, then there would be valid resolved combinations where the plugins would simply not do anything on desktop.

I think we have enough evidence at this point that `any` isn't the right approach, and that we should update the spec.
